# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

# Get path and print contents of directory
path = os.getcwd()
print(path)
print(os.listdir(path))

/Users/tiffanyflor/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pandas/compat/__init__.py:117: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


/Users/tiffanyflor/Dropbox/MyProjects/GuidedCapstone
['updated_ski_data.csv', 'Notebook_step4and5.ipynb', '.DS_Store', 'GuidedCapstoneStep3HL.ipynb', 'models', 'GuidedCapstone_Step2HL.ipynb', 'GuidedCapstoneStep6HL.ipynb', 'figures', '.ipynb_checkpoints', 'data']


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
# Using Google Colab
#path_to_data = 'step3_output.csv'

# Using computer with data folder
path_to_data = 'data/step3_output.csv'

# Create dataframe of step 3 output using path_to_data
df = pd.read_csv(path_to_data)

df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         1  
4          78.0              104.0            80.0         1  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
# Select categorical column state, creating a data frame subset named dfo
dfo = df[['state']]
dfo.head()

state
0   Alaska
1   Alaska
2   Alaska
3  Arizona
4  Arizona

In [4]:
# Concatenate the original data frame df while dropping column selected in dfo with get_dummies command
dummy = pd.get_dummies(dfo)
df.drop(dfo,axis=1,inplace=True)
df = pd.concat([df, dummy], axis=1)
df.head()


Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  state_Rhode Island  \
0        0.0          0          2     2       0  ...                   0   
1        0.0          0          0     0       0  ...                   0   
2        0.0          0          0     0       1  ...                   0   
3        0.0          1          0     2       2  ...                   0   
4        0.0          0          1     2       3  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we don't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [5]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [6]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [7]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [8]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [9]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [10]:
explained_variance_score(y_test, y_pred)

0.934716312708213

In [11]:
mean_absolute_error(y_test, y_pred)

5.054493992641167

In [12]:
from sklearn.metrics import mean_squared_error
from math import sqrt
sqrt(mean_squared_error(y_test, y_pred))

6.678369752581653

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
print(lm.intercept_)

64.10584668705994


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)


Coefficient
total_chairs          1.865907e+13
fastQuads             7.073696e+12
surface               6.627523e+12
state_New York        6.453687e+12
state_Michigan        6.090528e+12
double                5.840419e+12
state_Colorado        5.366108e+12
state_California      5.251237e+12
triple                5.210055e+12
state_Pennsylvania    5.011061e+12
state_Wisconsin       4.620591e+12
state_New Hampshire   4.620591e+12
state_Vermont         4.480986e+12
state_Minnesota       4.335910e+12
quad                  4.222557e+12
state_Utah            4.184793e+12
state_Montana         4.026956e+12
state_Idaho           4.026956e+12
state_Massachusetts   3.861574e+12
state_Oregon          3.687633e+12
state_Washington      3.687633e+12
state_New Mexico      3.503858e+12
state_Maine           3.503858e+12
state_Wyoming         3.308610e+12
state_North Carolina  2.874225e+12
state_Ohio            2.627843e+12
state_Connecticut     2.627843e+12
state_Nevada          2.354027e+12
state_Illinois        2.354027e+12
state_Virginia        2.354027e+12
state_West Virginia   2.354027e+12
fastSixes             2.097001e+12
state_Iowa            2.041772e+12
state_Alaska          2.041772e+12
trams                 1.801801e+12
state_South Dakota    1.669647e+12
state_Arizona         1.669647e+12
state_Indiana         1.669647e+12
state_Missouri        1.669647e+12
state_New Jersey      1.669647e+12
state_Rhode Island    1.182417e+12
state_Maryland        1.182417e+12
state_Tennessee       1.182417e+12
fastEight             1.771347e+11
AdultWeekday          1.919353e+01
summit_elev           1.822471e+01
base_elev             1.717798e+01
vertical_drop         5.652944e+00
NightSkiing_ac        7.692028e-01
Runs                  6.055870e-01
daysOpenLastYear      5.224311e-01
Snow Making_ac        4.960109e-01
LongestRun_mi         4.368156e-01
yearsOpen             3.786168e-01
clusters              3.681281e-01
TerrainParks          1.804096e-01
SkiableTerrain_ac     1.587460e-01
projectedDaysOpen     1.331243e-01
averageSnowfall       1.151295e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [21]:
# PREP TO CREATE MODEL2
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
drop_me = ([col for col in df.columns if 'state_' in col]) + ['Name','AdultWeekend']
X = df.drop(drop_me, axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)


In [22]:
#all second model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

print('Explained Variance Score: ' + str(explained_variance_score(y_test, y_pred)))
print('Mean Absolute Error: ' + str(mean_absolute_error(y_test, y_pred)))
print('Intercept: ' + str(lm.intercept_))

Explained Variance Score: 0.9257281142134342
Mean Absolute Error: 5.4277642993080875
Intercept: 64.06239374707621


In [23]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         19.927404
summit_elev          13.349243
base_elev            11.293093
vertical_drop         4.885929
averageSnowfall       1.689956
quad                  1.556640
triple                1.421926
surface               1.353837
clusters              1.295898
Runs                  1.274865
daysOpenLastYear      1.150176
fastQuads             0.683006
total_chairs          0.643580
NightSkiing_ac        0.582438
fastSixes             0.395869
projectedDaysOpen     0.367608
double                0.304942
yearsOpen             0.272230
SkiableTerrain_ac     0.257257
LongestRun_mi         0.239967
trams                 0.237443
TerrainParks          0.215881
fastEight             0.083507
Snow Making_ac        0.051236

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [24]:
# PREP TO CREATE MODEL3
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
drop_me = drop_me + ['summit_elev','base_elev']
X = df.drop(drop_me, axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)


In [25]:
#all third model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

print('Explained Variance Score: ' + str(explained_variance_score(y_test, y_pred)))
print('Mean Absolute Error: ' + str(mean_absolute_error(y_test, y_pred)))
print('Intercept: ' + str(lm.intercept_))

Explained Variance Score: 0.9313136899657813
Mean Absolute Error: 5.237370092894817
Intercept: 64.06402559903322


In [26]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         19.902847
vertical_drop         1.680613
averageSnowfall       1.621054
quad                  1.512615
triple                1.403843
Runs                  1.382259
surface               1.289715
clusters              1.259339
daysOpenLastYear      1.222912
fastQuads             0.678088
total_chairs          0.650290
NightSkiing_ac        0.614522
fastSixes             0.439517
projectedDaysOpen     0.349919
double                0.300000
trams                 0.275381
yearsOpen             0.268706
SkiableTerrain_ac     0.256341
LongestRun_mi         0.229405
TerrainParks          0.158532
Snow Making_ac        0.139377
fastEight             0.034003

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.935 | 5.05 |-|
| Model 2. | 0.926| 5.42 |'state'|
| Model 3. | 0.931 | 5.24 |'state','summit_elev','base_elev'|

Model Selection: #3